Modules/_weakref.c

```diff
@@ -86,26 +82,57 @@ Return a list of all weak reference objects pointing to 'object'.
 [clinic start generated code]*/
 
 static PyObject *
-_weakref_getweakrefs_impl(PyObject *module, PyObject *object)
-/*[clinic end generated code: output=5ec268989fb8f035 input=3dea95b8f5b31bbb]*/
+_weakref_getweakrefs(PyObject *module, PyObject *object)
+/*[clinic end generated code: output=25c7731d8e011824 input=00c6d0e5d3206693]*/
 {
     if (!_PyType_SUPPORTS_WEAKREFS(Py_TYPE(object))) {
         return PyList_New(0);
     }
 
     PyWeakReference **list = GET_WEAKREFS_LISTPTR(object);
-    Py_ssize_t count = _PyWeakref_GetWeakrefCount(*list);
+    Py_ssize_t count = _PyWeakref_GetWeakrefCountThreadsafe(object);
 
     PyObject *result = PyList_New(count);
     if (result == NULL) {
         return NULL;
     }
 
+#ifdef Py_GIL_DISABLED
```

Comment:
This seems more complicated than necessary. Don't bother with the count, just call `PyList_Append` within the loop with the lock held.

Both the default build and the free-threaded build need to handle weakrefs with zero refcount that are not yet removed from the list. Let's add a variant of `_Py_TryIncref` that works with both builds and use it here. Something like:

```
_Py_TryIncref(PyObject *op)
{
#ifdef Py_GIL_DISABLED
 ...
#else
if (Py_REFNT(op) > 0) {
    Py_INCREF(op);
    return 1;
}
return 0;
}
```

In other words, let's have one loop for both builds and avoid the list trimming logic.

Reply:
I *think* that's safe... my concern is that we aren't using a critical section to lock the weakref lock, so we potentially risk deadlock by doing so.